diff --git a/Examples_old/ROS/ORB_SLAM3/src/AR/ViewerAR.cc b/Examples_old/ROS/ORB_SLAM3/src/AR/ViewerAR.cc
index dc9f18b..163ac1f 100644
--- a/Examples_old/ROS/ORB_SLAM3/src/AR/ViewerAR.cc
+++ b/Examples_old/ROS/ORB_SLAM3/src/AR/ViewerAR.cc
@@ -17,6 +17,7 @@
 */
 
 #include "ViewerAR.h"
+#include "../../../include/Converter.h"
 
 #include <opencv2/highgui/highgui.hpp>
 
@@ -402,7 +403,7 @@ Plane* ViewerAR::DetectPlane(const cv::Mat Tcw, const std::vector<MapPoint*> &vM
         {
             if(pMP->Observations()>5)
             {
-                vPoints.push_back(pMP->GetWorldPos());
+		vPoints.push_back(ORB_SLAM3::Converter::toCvMat(pMP->GetWorldPos()));
                 vPointMP.push_back(pMP);
             }
         }
@@ -527,7 +528,7 @@ void Plane::Recompute()
         MapPoint* pMP = mvMPs[i];
         if(!pMP->isBad())
         {
-            cv::Mat Xw = pMP->GetWorldPos();
+	    cv::Mat Xw = ORB_SLAM3::Converter::toCvMat(pMP->GetWorldPos());
             o+=Xw;
             A.row(nPoints).colRange(0,3) = Xw.t();
             nPoints++;
